beacon-chain/blockchain/service.go

```diff
@@ -144,12 +144,16 @@ func (s *Service) Start() {
 func (s *Service) Stop() error {
 	defer s.cancel()
 
+	// lock before accessing s.head, s.head.state, s.head.state.FinalizedCheckpoint().Root
+	s.headLock.RLock()
```

Comment:
Same as above, one option to reduce the lock time
```go
s.headLock.RLock()
r := s.head.state.FinalizedCheckpoint().Root
s.headLock.RUnlock()
s.cfg.StateGen.ForceCheckpoint(s.ctx, r)
```

Reply:
also update in ac7c9b794acd6ca1e38ca11f0dab19e559015283

Reply:
My bad. My example above was not great. It defeats the nil checks. I think 👇🏼  makes the most sense
```go
	s.headLock.RLock()
	if s.cfg.StateGen != nil && s.head != nil && s.head.state != nil {
		r := s.head.state.FinalizedCheckpoint().Root
		s.headLock.RUnlock()
		// Save the last finalized state so that starting up in the following run will be much faster.
		if err := s.cfg.StateGen.ForceCheckpoint(s.ctx, r); err != nil {
			return err
		}
	}
	s.headLock.RUnlock()
```

Reply:
This makes sense to me. Updated.